force-app/main/default/classes/DonationHistoryController_TEST.cls

```diff
@@ -0,0 +1,120 @@
+/*
+    Copyright (c) 2021, Salesforce.org
+    All rights reserved.
+    
+    Redistribution and use in source and binary forms, with or without
+    modification, are permitted provided that the following conditions are met:
+    
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    * Neither the name of Salesforce.org nor the names of
+      its contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+    "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
+    LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
+    FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
+    COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
+    INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
+    BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
+    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER 
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT 
+    LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN 
+    ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
+    POSSIBILITY OF SUCH DAMAGE.
+*/
+/**
+* @author Salesforce.org
+* @date 2021
+* @group Donation
+* @description Class for test list donation history controller.
+*/
+@IsTest
+public class DonationHistoryController_TEST {
+    @IsTest
+    public static void testGetDonationHistory() {
+        
+        Date myDateTime = Date.newInstance(2020, 2, 2);
+
+        Account mockDefaultAccount = new Account(Name = 'Default Account');
```

Comment:
Please review the `UTIL_UnitTestData_TEST` class for pre-existing test data factory methods.

Reply:
For this case we generate different opportunities and each one with a different configuration to be able to have all the types of credits, I did not find a method that allows us to make these configurations in the class, what exactly would you be suggesting in this case?

Reply:
There's certainly methods for generating Accounts and Contacts though.

Reply:
Yes I saw them, but we have methods like `getContact` that return a contact but without an ID (since it does not make the dml) which we need to reference in the creation of opportunities, also `createAccountContactTestData` which looks very useful but at the same time it does not return the accounts and contacts generated, it is also `oppsForContactListByRecTypeId` but do not reference the account id in the opportunity, since we need certain references, I just wanted to clarify what would be the suggestion to implement the methods of that class or in any case which one

Reply:
I'm sorry if this is generating confusion but since I am not aware of the resources, maybe I need a hand, thanks in advance @force2b

Reply:
>it does not return the accounts and contacts

This is a great observation! I think that `createAccountContactTestData` can be updated to return either the Contacts or Accounts that it creates. If I had to choose, I'd say that method should return a `List<Contact>`. Then you can build out the base data like this:

```
List<Contact> contacts = UTIL_UnitTestData_TEST.createAccountContactTestData(CAO_Constants.HH_ACCOUNT_TYPE, 1, 2, 0);
List<Opportunity> opportunities = UTIL_UnitTestData_TEST.oppsForContactList(contacts, null, 'Closed Won', myDateTime, 250, null, null);
insert opportunities;

```

There's no UTIL_UnitTestData_TEST method for creating OpportunityContactRoles, yet. I think one could be created here. A function with this signature should work:

`public static List<OpportunityContactRole> createRolesForOpportunities(List<Id> opportunityIds, Id contactId, String role)`

Reply:
Very good suggestions @bdvorachek. Overall though, I think what's missing in this test class is a solid set of test scenarios (see [yesterday's comment](https://github.com/SalesforceFoundation/NPSP/pull/6760#discussion_r773126218)). This test class has one test method that tests only one possibility. There are way more scenarios to test. Plus, I'm questioning whether DML is even needed to fully test this logic. The service class could have the selector results injected with mock data. That would open this test class up to being able to test multiple scenarios very quickly.

Reply:
added test cases to check permissions and no results